### PR TITLE
feat(gas): do not store self-code in dictionary of child contract codes

### LINKED ISSUE #1012: Do not store self smart-contract code: use MYCODE asm instruction instead
## The problem
Lets take a look on a 2 simple smart-contracts:

```tact
contract Parent{
    id: Int = 0;
    init(){}
    receive("new"){
        self.id += 1;
        let data = initOf Child(self.id);
        self.forward(contractAddress(data), "init".asComment(), true, data);
    }
}

contract Child{
    id: Int;
    init(id: Int){
        self.id = id;
    }

    receive("init"){}

    receive("deploy_random"){
        let data = initOf Child(randomInt());
        self.forward(contractAddress(data), "init".asComment(), true, data);
    }
}
```

Under the hood, both Parent and Child contracts stores the code of `Child` smart contract in the system dictionary, because  in both of them `initOf Child()` used.

But storing code of `Child` smart contract inside `Child` is useless because one can call `MYCODE` [asm instruction](https://docs.ton.org/v3/documentation/tvm/changelog/tvm-upgrade-2023-07#opcodes-to-work-with-new-c7-values) to get it.

This approach would reduce not only storageFees, but also forwardFees when initializing new smart-contracts. 

It is really useful feature for developing `Jetton` smart contracts (when one child deploys another child), which optimizes fees a lot.

Feel free to argue to my point.

### PR DESCRIPTION
**Pull request by anton-trunov**
This should reduce forward fees for Jettons, for instance.

Another possible optimization would be to remove the child contract dictionary for the common case when there are just one or very few child contracts and use cell references to store the child contract codes. But let's do it as a separate PR so we can check its impact too.

## Issue

Closes #1012.

## Checklist

- [x] I have updated CHANGELOG.md
- [x] I have added tests to demonstrate the contribution is correctly implemented: it reduces gas consumption for the common cases, makes storage improvements, etc. and does not have too much negative impact for the rest of the use-cases
- [x] I have run all the tests locally and no test failure was reported
- [x] I have run the linter, formatter and spellchecker
- [x] I did not do unrelated and/or undiscussed refactorings

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -27,6 +27,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Do not automatically validate all addresses when receiving/sending messages or using address manipulating functions: PR [#1207](https://github.com/tact-lang/tact/pull/1207)
 - Remove `enabledMasterchain` compiler config option from `tact.config.json`: PR [#1207](https://github.com/tact-lang/tact/pull/1207)
 - Remove `org.ton.chain.any.v0` interface: PR [#1207](https://github.com/tact-lang/tact/pull/1207)
+- Tact does not anymore store the code of the parent contract in the system cell with all the child contract codes used in `initOf` to reduce gas fees: PR [#1213](https://github.com/tact-lang/tact/pull/1213)
```
Comment (novusnota) [inline CHANGELOG.md]:
```suggestion
- To reduce fees, Tact no longer stores the parent contract code in the system cell that holds all the child contract codes used in `initOf`. Instead, the `MYCODE` instruction is used: PR [#1213](https://github.com/tact-lang/tact/pull/1213)
```

Readability nitpick

Comment (anton-trunov) [reply CHANGELOG.md]: `s/held/holds`

**Review by Gusarich — DISMISSED** [review]
Looks good, but for some reason it increases fees for some transactions in `initof` tests 🤔 we definetely need to benchmark this with https://github.com/tact-lang/jetton

**Comment by anton-trunov** [comment]
> for some reason it increases fees for some transactions in initof tests

do you still have the exact numbers?

because it definitely decreased gas in the benchmarks: image

**Comment by anton-trunov** [comment]
Nice results! Let's merge

image

**Pull request merged**